### PR TITLE
Implement a new shared "Drainer" handler.

### PR DESCRIPTION
**Pull request by mattmoor**
This implements a new `http.Handler` called `Drainer`, which is intended to wrap some inner `http.Handler` business logic with a new outer handler that can respond to Kubelet probes (successfully until told to "Drain()").

This takes over the webhook's relatively new probe handling and lame duck logic with one key difference.  Previously the webhook waited for a fixed period after SIGTERM before exitting, but the new logic waits for this same grace period AFTER THE LAST REQUEST.  So if the handler keeps getting (non-probe) requests, the timer will continually reset, and once it stops receiving requests for the configured grace period, "Drain()" will return and the webhook will exit.

The goal of this work is to try to better cope with what we believe to be high tail latencies of the API server seeing that a webhook replica is shutting down.

Related: https://github.com/knative/pkg/issues/1509

### CODE REVIEW

File: network/handlers/drain.go [hunk 1]
```diff
@@ -0,0 +1,105 @@
...
+	// Mutex guards the initialization and resets of the timer
+	sync.Mutex
+
+	// Inner is the http.Handler to which we delegate actual requests.
+	Inner http.Handler
+
+	// QuietPeriod is the duration that must elapse without any requests
+	// after Drain is called before it may return.
+	QuietPeriod time.Duration
+
```
Comment (vagababov) [inline network/handlers/drain.go:43]: I'd hide those as well and add a ctor

File: network/handlers/drain.go [hunk 2]
```diff
@@ -0,0 +1,105 @@
...
+}
+
+// Ensure Drainer implements http.Handler
+var _ http.Handler = (*Drainer)(nil)
+
+// ServeHTTP implements http.Handler
+func (d *Drainer) ServeHTTP(w http.ResponseWriter, r *http.Request) {
+	if network.IsKubeletProbe(r) { // Respond to probes regardless of path.
+		if d.draining() {
+			http.Error(w, "shutting down", http.StatusInternalServerError)
```
Comment (vagababov) [inline network/handlers/drain.go:58]: 503?

File: network/handlers/drain.go [hunk 3]
```diff
@@ -0,0 +1,105 @@
...
+
+// Drainer wraps an inner http.Handler to support responding to kubelet
+// probes with a "200 OK" until the handler is told to Drain.
+// When the Drainer is told to Drain, it will immediately start to fail
+// probes with a "500 shutting down", and the call will block until no
+// requests have been received for QuietPeriod (defaults to
+// network.DefaultDrainTimeout).
+type Drainer struct {
+	// Mutex guards the initialization and resets of the timer
+	sync.Mutex
```
Comment (vagababov) [inline network/handlers/drain.go:35]:
```suggestion
	sync.RWMutex
```

File: network/handlers/drain.go [hunk 4]
```diff
@@ -0,0 +1,105 @@
...
+	}
+	if d.timer.Stop() {
+		d.timer.Reset(d.QuietPeriod)
+	}
+}
+
+// draining returns whether we are draining the handler.
+func (d *Drainer) draining() bool {
+	d.Lock()
+	defer d.Unlock()
```
Comment (vagababov) [inline network/handlers/drain.go:102]:
```suggestion
	d.RLock()
	defer d.RUnlock()
```
With RLock is definitely better, than the write lock above, but still atomic would probably be faster.

File: network/handlers/drain.go [hunk 5]
```diff
@@ -0,0 +1,105 @@
...
+}
+
+// Drain blocks until QuietPeriod has elapsed since the last request,
+// starting when this is invoked.
+func (d *Drainer) Drain() {
+	d.once.Do(func() {
+		t := func() *time.Timer {
+			d.Lock()
+			defer d.Unlock()
+			if d.QuietPeriod == 0 {
```
Comment (vagababov) [inline network/handlers/drain.go:76]:
```suggestion
			if d.QuietPeriod <= 0 {
```

File: network/handlers/drain.go [hunk 6]
```diff
@@ -0,0 +1,105 @@
...
+	})
+}
+
+// reset resets the drain timer to the full amount of time.
+func (d *Drainer) reset() {
+	d.Lock()
+	defer d.Unlock()
+	if d.timer == nil {
+		return
+	}
```
Comment (vagababov) [inline network/handlers/drain.go:93]: On every request lock/unlock is basically serializing everything.
Probably not the best idea.
You can probably replace it with a atomic 0/1 toggle, that's much cheaper to check.

File: network/handlers/drain.go (hunk 3 above)
Comment (mattmoor) [reply network/handlers/drain.go]: done

File: network/handlers/drain.go (hunk 1 above)
Comment (mattmoor) [reply network/handlers/drain.go:43]: Makes it more complicated to have an optional QuietPeriod 🤷

File: network/handlers/drain.go (hunk 6 above)
Comment (mattmoor) [reply network/handlers/drain.go]: With the tweaks I added it should be exclusively using the read locks until Drain is called, which should be cheap enough.  It also only wraps part of the request.

File: network/handlers/drain.go (hunk 4 above)
Comment (mattmoor) [reply network/handlers/drain.go]: done

File: network/handlers/drain.go (hunk 1 above)
Comment (vagababov) [reply network/handlers/drain.go:43]: just pass 0 as QP? ;-)

Comment (mattmoor) [reply network/handlers/drain.go:43]: 🤢

**Pull request merged**